File: aoc/2021/day-3/2.py
import fileinput
lines = [i.strip() for i in fileinput.input("aoc/2021/day-3/input.txt")]
oxygen = 0
co2 = 0

copy = lines.copy()
for i in range(len(lines[0])):
  ones, zeros = 0, 0
  for line in copy:
    if line[i] == '1':
      ones += 1
    else:
      zeros += 1
  print(copy)
  copy = list(filter(lambda x: x[i] == "1" if ones >= zeros else "0", copy))
  if len(copy) == 1:
    oxygen = int(copy[0], 2)
    break

copy = lines.copy()
for i in range(len(lines[0])):
  ones, zeros = 0, 0
  for line in copy:
    if line[i] == '1':
      ones += 1
    else:
      zeros += 1
  copy = list(filter(lambda x: x[i] == "0" if ones > zeros else "1", copy))
  if len(copy) == 1:
    co2 = int(copy[0], 2)
    break

print(oxygen * co2)

File: aoc/2021/day-3/2.js
// read file input.txt and split on newline
const { join } = require("path");
let oxygen = 0,
  co2 = 0;
const input = require("fs")
  .readFileSync(join(__dirname, "./input.txt"), "utf8")
  .split("\n");
let copy = [...input];
for (let i = 0; i < input[0].length; i++) {
  let ones = 0,
    zeros = 0;
  for (const line of copy) {
    if (line[i] === "1") ones++;
    else zeros++;
  }
  copy = copy.filter((x) => (ones >= zeros ? x[i] === "1" : x[i] === "0"));
  if (copy.length === 1) {
    oxygen = parseInt(copy[0], 2);
    break;
  }
}

copy = [...input];
for (let i = 0; i < input[0].length; i++) {
  let ones = 0,
    zeros = 0;
  for (const line of copy) {
    if (line[i] === "1") ones++;
    else zeros++;
  }
  copy = copy.filter((x) => (ones >= zeros ? x[i] === "0" : x[i] === "1"));
  if (copy.length === 1) {
    co2 = parseInt(copy[0], 2);
    break;
  }
}

console.log(oxygen * co2);
